File: examples_rclpy/examples_rclpy/subscriber/observer.py
#!/usr/bin/env python3
#
# Copyright 2019 ROBOTIS CO., LTD.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Authors: Darby Lim, Pyo

from rcl_interfaces.msg import ParameterDescriptor
from rcl_interfaces.msg import ParameterEvent
from rclpy.node import Node
from rclpy.parameter import Parameter
from rclpy.qos import QoSProfile
from rclpy.qos import qos_profile_action_status_default
from rclpy.qos import qos_profile_parameter_events
from rclpy.qos import qos_profile_parameters
from rclpy.qos import qos_profile_sensor_data
from rclpy.qos import qos_profile_services_default
from rclpy.qos import qos_profile_system_default

from examples_msgs.msg import Count


class Observer(Node):

    def __init__(self, qos_profile):
        super().__init__('observer')

        self.get_logger().debug('Test debug message')

        self.init_parameters()
        self.parameter_event_callback()

        self.i = 0
        self.blind_count = 0

        qos = self.get_qos(qos_profile)
        self.sub = self.create_subscription(Count, 'count', self.counter_callback, qos)

    def counter_callback(self, msg):
        self.i += 1

        if self.blind:
            if self.blind_count >= 10:
                self.set_parameters([Parameter('blind', Parameter.Type.BOOL, False)])
                self.blind = self.get_parameter('blind').value
                self.blind_count = 0

            self.get_logger().info('Blind mode')
            self.blind_count += 1
        else:
            self.get_logger().info(
                '[{0}] Observed: \'{1}\''.format(self.comment, msg.count + self.offset))

    def param_event_callback(self, event):
        for new_parameter in event.new_parameters:
            self.get_logger().info('New parameter name : {0}'.format(new_parameter.name))

        for changed_parameter in event.changed_parameters:
            self.get_logger().info('Changed parameter name : {0}'.format(changed_parameter.name))

            if changed_parameter.name == 'blind':
                self.blind = Parameter.from_parameter_msg(changed_parameter).value
            elif changed_parameter.name == 'offset':
                self.offset = Parameter.from_parameter_msg(changed_parameter).value
            elif changed_parameter.name == 'comment':
                self.comment = Parameter.from_parameter_msg(changed_parameter).value

        for deleted_parameter in event.deleted_parameters:
            self.get_logger().info('Deleted parameter name : {0}'.format(deleted_parameter.name))

    def init_parameters(self):
        self.declare_parameter('comment', 'No comments', ParameterDescriptor())
        self.declare_parameter('blind', False, ParameterDescriptor())
        self.declare_parameter('offset', 0, ParameterDescriptor())
        self.declare_parameter('favorite/numbers_integers', [1, 2], ParameterDescriptor())
        self.declare_parameter('favorite/numbers_doubles', [1.0, 2.0], ParameterDescriptor())

        self.comment = self.get_parameter('comment').value
        self.blind = self.get_parameter('blind').value
        self.offset = self.get_parameter('offset').value

    def parameter_event_callback(self):
        self.param_event_sub = self.create_subscription(
            ParameterEvent,
            'parameter_events',
            self.param_event_callback,
            qos_profile_parameter_events)

    def get_qos(self, i):
        return {
            0: QoSProfile(depth=10),
            1: qos_profile_sensor_data,
            2: qos_profile_parameters,
            3: qos_profile_services_default,
            4: qos_profile_parameter_events,
            5: qos_profile_system_default,
            6: qos_profile_action_status_default}[i]

File: examples_rclcpp/src/subscriber/observer.cpp
/*******************************************************************************
* Copyright 2019 ROBOTIS CO., LTD.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/* Authors: Darby Lim, Pyo */

#include "subscriber/observer.hpp"

using namespace robotis;

// https://en.cppreference.com/w/cpp/symbol_index/chrono_literals
using namespace std::chrono_literals;

Observer::Observer(const int32_t & qos_profile)
: Node("observer")
{
  RCLCPP_DEBUG(this->get_logger(), "Test debug message");

  init_parameters();
  parameter_event_callback();

  // Set QoS profile
  auto qos = get_qos(qos_profile);

  // Rambda function (https://en.cppreference.com/w/cpp/language/lambda)
  // 'typename' specify that this is a type
  sub_ = this->create_subscription<examples_msgs::msg::Count>(
    "count",
    qos,
    [this](const typename examples_msgs::msg::Count::SharedPtr msg) -> void
      {
        static uint8_t blind_count = 0;

        if (blind_)
        {
          if (blind_count >= 10)
          {
            this->set_parameters({rclcpp::Parameter("blind", false)});
            this->get_parameter("blind", blind_);

            blind_count = 0;
          }

          RCLCPP_INFO(this->get_logger(), "Blind mode");

          blind_count++;
        }
        else
        {
          RCLCPP_INFO(
            this->get_logger(),
            "[%s] Observed: '%d'",
            comment_.c_str(), msg->count + offset_);
        }
      }
    );
}

void Observer::init_parameters()
{
  // Declare parameters that may be set on this node
  this->declare_parameter(
    "comment",
    rclcpp::ParameterValue("No comments"),
    rcl_interfaces::msg::ParameterDescriptor());
  this->declare_parameter(
    "blind",
    rclcpp::ParameterValue(false),
    rcl_interfaces::msg::ParameterDescriptor());
  this->declare_parameter(
    "offset",
    rclcpp::ParameterValue(0),
    rcl_interfaces::msg::ParameterDescriptor());
  this->declare_parameter(
    "favorite.numbers_integers",
    rclcpp::ParameterValue(std::vector<int64_t>({1, 2})),
    rcl_interfaces::msg::ParameterDescriptor());
  this->declare_parameter(
    "favorite.numbers_doubles",
    rclcpp::ParameterValue(std::vector<double>({1.0, 2.0})),
    rcl_interfaces::msg::ParameterDescriptor());

  // Get parameter from yaml
  this->get_parameter("comment", comment_);
  this->get_parameter("blind", blind_);
  this->get_parameter("offset", offset_);

  std::vector<int64_t> int_nums =
    this->get_parameter("favorite.numbers_integers").as_integer_array();
  std::vector<double> double_nums =
    this->get_parameter("favorite.numbers_doubles").as_double_array();
}

void Observer::parameter_event_callback()
{
  parameters_client_ = std::make_shared<rclcpp::AsyncParametersClient>(this);
  while (!parameters_client_->wait_for_service(1s))
  {
    if (!rclcpp::ok())
    {
      RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
      return;
    }
    RCLCPP_INFO(this->get_logger(), "service not available, waiting again...");
  }

  // Setup callback for changes to parameters from parameter server.
  // (ref: http://docs.ros2.org/dashing/api/rclcpp/classrclcpp_1_1Parameter.html)
  auto param_event_callback =
    [this](const rcl_interfaces::msg::ParameterEvent::SharedPtr event) -> void
      {
        for (auto & new_parameter : event->new_parameters)
        {
          RCLCPP_INFO(this->get_logger(), "new parameter name : %s", new_parameter.name.c_str());
        }

        for (auto & changed_parameter : event->changed_parameters)
        {
          RCLCPP_INFO(
            this->get_logger(),
            "changed parameter name : %s",
            changed_parameter.name.c_str());

          if (changed_parameter.name == "blind")
          {
            auto value = rclcpp::Parameter::from_parameter_msg(changed_parameter).as_bool();
            blind_ = value;
          }
          else if (changed_parameter.name == "offset")
          {
            auto value = rclcpp::Parameter::from_parameter_msg(changed_parameter).as_int();
            offset_ = value;
          }
          else if (changed_parameter.name == "comment")
          {
            auto value = rclcpp::Parameter::from_parameter_msg(changed_parameter).as_string();
            comment_ = value;
          }
        }

        for (auto & deleted_parameter : event->deleted_parameters)
        {
          RCLCPP_INFO(
            this->get_logger(),
            "deleted parameter name : %s",
            deleted_parameter.name.c_str());
        }
      };

  parameter_event_sub_ = parameters_client_->on_parameter_event(param_event_callback);
}

rclcpp::QoS Observer::get_qos(const int32_t & qos_profile)
{
  auto qos = rclcpp::QoS(rclcpp::KeepLast(10));

  if (qos_profile == 0)
  {
    qos = qos;
  }
  else if (qos_profile == 1)
  {
    qos = rclcpp::QoS(rclcpp::SensorDataQoS());
  }
  else if (qos_profile == 2)
  {
    qos = rclcpp::QoS(rclcpp::ParametersQoS());
  }
  else if (qos_profile == 3)
  {
    qos = rclcpp::QoS(rclcpp::ServicesQoS());
  }
  else if (qos_profile == 4)
  {
    qos = rclcpp::QoS(rclcpp::ParameterEventsQoS());
  }
  else if (qos_profile == 5)
  {
    qos = rclcpp::QoS(rclcpp::SystemDefaultsQoS());
  }
  else
  {
    qos = qos;
  }

  return qos;
}
